### PR TITLE
Team guidelines for samples

### PR DESCRIPTION
**Pull request by qwwdfsad**

### CODE REVIEW

File: team_docs/samples.md [hunk 1]
```diff
@@ -0,0 +1,239 @@
+
+Guidelines for samples in documentation 
+--------------------------------------------------------
+The guideline provides a skeleton of what substitutes a good sample to the Kotlin signature. These are general recommendations, not strict rules. If ignoring some of the recommendations improves your documentation -- do that.
```
Comment (ilya-g) [inline team_docs/samples.md:4]: "Kotlin signature" - I believe it's very unusual to use this term to describe functions, properties and classifiers, better to use more established "Kotlin declaration"

File: team_docs/samples.md [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+### Classes
+
+A class consists of three parts:
+
+* How its objects are obtained;
+* How its objects are used;
+* What other entities are available in its namespace (objects, functions, etc.).
+
+It is recommended providing top-level overview in the documentation for each of the aspect, ideally with the help of inline examples
+#### Construction
```
Comment (ilya-g) [inline team_docs/samples.md:166]: Consistent spacing between/after headings

File: team_docs/samples.md [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+"extra" behavior, whereas member functions encode the core behavior.
+
+The top-level documentation should include an overview of the core behavior
+(typically, of the member functions), with simplified examples of each to
+quickly show one what behaviors are available. In these brief overviews,
+the rules for providing samples to functions don't need to be followed;
+instead, the sample must only demonstrate how *some* inputs and *some* forms of
+the output are linked. Edge cases and uncommon paths can be ignored.
+However, the rough idea of what the function does must still be deducible.
+
```
Comment (dkhalanskyjb) [inline team_docs/samples.md:194]: #### Properties and fields

Properties and fields should be treated just like functions that accept the
class containing them as an argument; so, `val size: Int` of class `List`
should be accompanied by the samples that would be written for
`fun List<*>.size(): Int` according to the rules above:

```kotlin
check(listOf("a", "b", "c", "d").size == 4)
```

Mutable properties that have non-trivial setter behavior must also be treated
as functions that perform an action:

```kotlin
val input = "Mon, 30 Jul 2008 11:05:30 GMT"
val parsed = DateTimeComponents.Formats.RFC_1123.parse(input)
check(parsed.monthNumber == 7)
check(parsed.month == Month.JULY)
parsed.month = Month.JUNE
check(parsed.monthNumber == 6)
check(parsed.month == Month.JUNE)
```

File: team_docs/samples.md (hunk 1 above)
Comment (qwwdfsad) [reply team_docs/samples.md:4]: End up with "Kotlin API entry"

File: team_docs/samples.md (hunk 3 above)
Comment (qwwdfsad) [reply team_docs/samples.md:196]: I've adjusted your recommendation in a separate commit, PTAL.
Rationale:

* Got rid of the notion of `field`. These are non-existing (except for things like `@JvmField`)
* Got rid of a trivial example -- it both makes the text a little bit more compact without the loss of the general idea and no longer promotes a very basic and trivial sample (trivial samples might be okay, but should be picked with care)

File: team_docs/samples.md [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+    * Pros of asserts: explicitly asserting, very familiar, test-friendly (`@sample` is a test)
+    * Cons: when copy-pasted as is, the code is unresolbed, `main` rarely has asserts behaviour
+        * Seva: still can be ok as readability here trumps reusability?
+        * Also, can try to push converter to idea -- asserts to the main code -> checks
+    * Pros of `check`s: works out of the box, copy-pasteable, will work in scratches with a less implementation
+    * Cons: might be slightly less readable, as checks are less familiar and also do not encapsulate the intent (e.g. `check(sample.contains("foo"))` vs `assertContains(sample, "foo")`
+* Both `inline` code blocks for [context-heavy examples](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/callback-flow.html) and `@sample` tag can be used.
+* It might be reasonable to provide multiple samples for a single signature -- starting from the most generic one and going down on specifics (interface implementation, error handling, pitfalls workarounds)
+
+## Specific recommendations
```
Comment (qwwdfsad) [inline team_docs/samples.md:44]:
```suggestion

## Specific recommendations
```

File: team_docs/samples.md [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+### Functions returning a result
```
Comment (qwwdfsad) [inline team_docs/samples.md:45]:
```suggestion

### Functions returning a result
```

File: team_docs/samples.md (hunk 3 above)
Comment (qwwdfsad) [reply team_docs/samples.md:197]: Sorry, I haven't pushed it. I've meant the example with `size`. PTAL

Comment (dkhalanskyjb) [reply team_docs/samples.md:197]: Yeah, I, too, meant the example with `size`: I think that, for how clear `size` inherently is, a one-line example is sufficient.

Here's a longer sample for the same idea, if by trivial you mean that it's too short:
```kotlin
check(LocalDate(2024, 1, 1).dayOfYear == 1)
check(LocalDate(2024, 1, 5).dayOfYear == 5)
check(LocalDate(2024, 2, 4).dayOfYear == 35)
check(LocalDate(2024, 12, 31).dayOfYear == 366)
```

**Review by fzhinkin — APPROVED** [review]

**Pull request merged**